File: pkmnapi-api/src/requests/pokemon_movesets.rs
use rocket_okapi::JsonSchema;
use serde::Deserialize;

use crate::requests::base::BaseRequest;

pub type PokemonMovesetRequest =
    BaseRequest<PokemonMovesetRequestType, PokemonMovesetRequestAttributes>;

impl PokemonMovesetRequest {
    pub fn get_moveset(&self) -> Vec<u8> {
        self.data
            .attributes
            .moveset
            .iter()
            .map(|moveset| moveset._move.id)
            .collect()
    }
}

#[derive(Debug, Deserialize, JsonSchema)]
#[allow(non_camel_case_types)]
pub enum PokemonMovesetRequestType {
    pokemon_movesets,
}

#[derive(Debug, Deserialize, JsonSchema)]
pub struct PokemonMovesetRequestAttributes {
    pub moveset: Vec<PokemonMovesetRequestAttributesMoveset>,
}

#[derive(Debug, Deserialize, JsonSchema)]
pub struct PokemonMovesetRequestAttributesMoveset {
    #[serde(rename = "move")]
    pub _move: PokemonMovesetRequestAttributesMovesetMove,
}

#[derive(Debug, Deserialize, JsonSchema)]
pub struct PokemonMovesetRequestAttributesMovesetMove {
    #[serde(deserialize_with = "crate::utils::from_numeric_str")]
    pub id: u8,
}

File: clients/js/src/requests/pokemon_movesets.ts
import { BaseRequest } from './base';

export type PokemonMovesetRequest = BaseRequest<PokemonMovesetRequestType, PokemonMovesetRequestAttributes>;

enum PokemonMovesetRequestType {
    pokemon_movesets = 'pokemon_movesets',
}

export interface PokemonMovesetRequestAttributes {
    moveset: Array<PokemonMovesetRequestAttributesMoveset>;
}

export interface PokemonMovesetRequestAttributesMoveset {
    move: PokemonMovesetRequestAttributesMovesetMove;
}

export interface PokemonMovesetRequestAttributesMovesetMove {
    id: string;
}

export const PokemonMovesetRequestDefault = (attributes: PokemonMovesetRequestAttributes): PokemonMovesetRequest => {
    return {
        data: {
            type: PokemonMovesetRequestType.pokemon_movesets,
            attributes,
        },
    };
};
